### PR TITLE
remove VmWorkJob after adding a nic to a vm

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR intents to

Fixes: #5541 without regressing #5651

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [x] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

manual tested

#### scenario 1: "Parallel adding of multiple networks"
create five shared networks,
create a VM in one of those
run a command line like 
```
for net in \
  "5e998528-6886-4d23-aa20-d8c0c3dfd0d4"\
  "759a94ac-3c36-425c-b3ef-94a2ada9f8a1"\
  "a7ce3c15-8497-4e52-9530-1a1ebf2cb023"\
  "8da9d26a-c139-49f0-ae64-fd38b331d48d"
do
    cmk add nictovirtualmachine networkid=$net virtualmachineid=c22686c7-1ff9-4e66-afe5-348127ef7fe2 &
done
```
substituting the right IDs

#### scenarion 2: "Parallel adding the same network multiple times"
create two shared networks,
create a VM in one of those
run a command line like 
```
for net in \
  "5e998528-6886-4d23-aa20-d8c0c3dfd0d4"\
  "5e998528-6886-4d23-aa20-d8c0c3dfd0d4"\
  "5e998528-6886-4d23-aa20-d8c0c3dfd0d4"
do
    cmk add nictovirtualmachine networkid=$net virtualmachineid=c22686c7-1ff9-4e66-afe5-348127ef7fe2 &
done
```
substituting the right IDs

#### scenario 3: "re-adding a network"
create two shared networks,
create a VM in one of those
add and remove the second one multiple times.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
still needs testing @weizhouapache m but can you review?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
PR doesn't seem ready cc @DaanHoogland @weizhouapache @nvazquez @sureshanaparti

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java
Comment (rhtyd) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:4060]: Should there be a default return if none of the cases are matched?

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java
Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:4060]: there should actually be a throw. let me check

**Review by weizhouapache — APPROVED** [review]
LGTM

verified the issue #5651 is fixed.

**Review by rhtyd — APPROVED** [review]
Adding to Wei's LGTM

**Comment by weizhouapache** [comment]
@blueorangutan test matrix

**Comment by nvazquez** [comment]
@blueorangutan test matrix

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
note that travis job 14 is showing regressions. we can't merge this.
it assumes cone in #5541 which should be reverted to get 4.16 out.
cc @div8cn

**Comment by rhtyd** [comment]
Ping any update/news on this @DaanHoogland cc @sureshanaparti @weizhouapache

**Comment by DaanHoogland** [comment]
No @rhtyd , this will need addressing. The original fix caused a regression.

**Comment by sureshanaparti** [comment]
@DaanHoogland can you fix the conflicts. thanks.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Comment by DaanHoogland** [comment]
ran one of the failing tests and it passed, now re-running the entire suite.

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java
Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:5667]: @weizhouapache the regression that this caused was that the nic couldn't be re-added because the workjob remained in the DB. Are you sure removing it here isn't needed?

**Comment by sureshanaparti** [comment]
Hi @DaanHoogland is this PR ready?

**Comment by DaanHoogland** [comment]
@sureshanaparti no

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java
Comment (weizhouapache) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:5667]: @DaanHoogland 
this PR lists pending work jobs by vm id and also network uuid.
If you add multiple networks to a vm, cloudstack will create work jobs for all of them, and process the jobs.

Comment (weizhouapache) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:5667]: @DaanHoogland 
I suggest you to try to reproduce the issue #5499 without last commit.

ps: I  found that some vm work jobs are in `Filed` step even if job is finished.

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java
Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: yes @weizhouapache I noticed that too.
The scenario from #5499 doen't work, only two nics get added to the VM. I'll keep looking

**Comment by DaanHoogland** [comment]
@sureshanaparti @weizhouapache I finally weeded out the final crinch here, a job was joined twice leading to an SQL exception. It did work but I'll start another test round for both parallel adding and re-adding of a network.
@blueorangutan package

**Comment by DaanHoogland** [comment]
fully works for adding in parallel and re-adding. I did notice however that deleting in parallel sometimes doesn't work. We will need to revisit the general area of this functionality again in the future
@blueorangutan test matrix

**Comment by weizhouapache** [comment]
> fully works for adding in parallel and re-adding. I did notice however that deleting in parallel sometimes doesn't work. We will need to revisit the general area of this functionality again in the future @blueorangutan test matrix

great ! @DaanHoogland

**Comment by rohityadavcloud** [comment]
Is this ready for merging, or needs testing @DaanHoogland @weizhouapache ?

**Comment by DaanHoogland** [comment]
> Is this ready for merging, or needs testing @DaanHoogland @weizhouapache @sureshanaparti ?

I have tested it but I am (co-)author It has my lgtm ;)
I'll add a brief test description.

**Review by weizhouapache — APPROVED** [review]

**Review by sureshanaparti — APPROVED** [review]
code LGTM

**Pull request merged**
